File: Solutions/743. Network Delay Time/743. Network Delay Time.java
//Dijkstra
class Solution {
    public int networkDelayTime(int[][] times, int N, int K) {
        List<List<int[]>> graph=new ArrayList<>();
        for(int i=0;i<=N;++i)
            graph.add(new ArrayList<>());
        for(int[] time:times){
            graph.get(time[0]).add(new int[]{time[1],time[2]});
        } 
        boolean[] visited=new boolean[N+1];
        int[] d=new int[N+1];
        Arrays.fill(d,Integer.MAX_VALUE);
        d[0]=d[K]=0;
        PriorityQueue<int[]> pq=new PriorityQueue<>((a,b)->(a[0]-b[0]));
        pq.offer(new int[]{0,K});
        while(!pq.isEmpty()){
            int[] curr=pq.poll();
            int dis=curr[0],node=curr[1];
            if(visited[node])
                continue;          
            for(var to:graph.get(node)){
                int v=to[0],w=to[1];
                if(dis+w<d[v]){
                    d[v]=dis+w;
                    pq.offer(new int[]{d[v],v});
                } 
            }    
            visited[node]=true;
        }
        int res=Integer.MIN_VALUE;
        for(int i=0;i<d.length;++i)
            res=res<d[i]?d[i]:res;
        return res==Integer.MAX_VALUE?-1:res;
    }
}






//Bellman Ford
class Solution {
    public int networkDelayTime(int[][] times, int N, int K) {
        int[] dist=new int[N+1];
        for(int i=0;i<=N;++i)
            dist[i]=Integer.MAX_VALUE;
        dist[0]=dist[K]=0;
        for(int i=0;i<N;++i)
            for(var t:times){
                int u=t[0], v=t[1], w=t[2];
                if(dist[u]!=Integer.MAX_VALUE&&dist[v]>dist[u]+w)
                    dist[v]=dist[u]+w;
            }
        int res=0;
        for(int i=1;i<=N;++i){
            res=Math.max(res,dist[i]);
        }
        return res==Integer.MAX_VALUE?-1:res;
    }
}

File: Solutions/743. Network Delay Time/743. Network Delay Time.py
#Dijkstra
class Solution:
    def networkDelayTime(self, times: List[List[int]], N: int, K: int) -> int:
        vp,visited,d=[],[],[]
        for i in range(N+1):
            vp.append([])
            visited.append(False)
            d.append(sys.maxsize)
        for t in times:
            vp[t[0]].append((t[1],t[2]))
        d[0],d[K]=0,0
        pq=[]
        heapify(pq)
        heappush(pq,(0,K))
        while pq:
            c,u=heappop(pq)
            if visited[u]:
                continue
            for to in vp[u]:
                v,w=to
                if c+w<d[v]:
                    d[v]=c+w
                    heappush(pq,(d[v],v))
            visited[u]=True
        res=max(d)
        return res if res!=sys.maxsize else -1







#Bellman Ford
class Solution:
    def networkDelayTime(self, times: List[List[int]], N: int, K: int) -> int:
        dist=[]
        for i in range(N+1):
            dist.append(0x77777777)
        dist[0]=dist[K]=0
        for i in range(N):
            for _,t in enumerate(times):
                u,v,w=t[0],t[1],t[2]
                if dist[u]!=0x77777777 and dist[v]>dist[u]+w:
                    dist[v]=dist[u]+w
        res=max(dist)
        if res==0x77777777:
            return -1
        return res